Security/Guard/Authenticator/OAuth2Authenticator.php

```diff
@@ -0,0 +1,126 @@
+use Symfony\Component\Security\Guard\AuthenticatorInterface;
+use Trikoder\Bundle\OAuth2Bundle\Security\Authentication\Token\OAuth2Token;
+use Trikoder\Bundle\OAuth2Bundle\Security\Authentication\Token\OAuth2TokenFactory;
+use Trikoder\Bundle\OAuth2Bundle\Security\Exception\InsufficientScopesException;
+use Trikoder\Bundle\OAuth2Bundle\Security\User\NullUser;
+
+/**
+ * @author Yonel Ceruto <yonelceruto@gmail.com>
+ * @author Antonio J. García Lagar <aj@garcialagar.es>
+ */
+final class OAuth2Authenticator implements AuthenticatorInterface
+{
+    private $httpMessageFactory;
+    private $resourceServer;
+    private $oauth2TokenFactory;
+    private $psr7Request;
+
+    public function __construct(HttpMessageFactoryInterface $httpMessageFactory, ResourceServer $resourceServer, OAuth2TokenFactory $oauth2TokenFactory)
+    {
+        $this->httpMessageFactory = $httpMessageFactory;
+        $this->resourceServer = $resourceServer;
+        $this->oauth2TokenFactory = $oauth2TokenFactory;
+    }
+
+    public function start(Request $request, ?AuthenticationException $authException = null): Response
+    {
+        $exception = new UnauthorizedHttpException('Bearer');
+
+        return new Response('', $exception->getStatusCode(), $exception->getHeaders());
+    }
+
+    public function supports(Request $request): bool
+    {
+        return 0 === strpos($request->headers->get('Authorization', ''), 'Bearer ');
+    }
+
+    public function getCredentials(Request $request)
+    {
+        $psr7Request = $this->httpMessageFactory->createRequest($request);
+
+        try {
+            $this->psr7Request = $this->resourceServer->validateAuthenticatedRequest($psr7Request);
+        } catch (OAuthServerException $e) {
+            throw new AuthenticationException('The resource server rejected the request.', 0, $e);
+        }
+
+        return $this->psr7Request->getAttribute('oauth_user_id');
+    }
+
+    public function getUser($userIdentifier, UserProviderInterface $userProvider): UserInterface
+    {
+        return '' === $userIdentifier ? new NullUser() : $userProvider->loadUserByUsername($userIdentifier);
+    }
+
+    public function checkCredentials($token, UserInterface $user): bool
+    {
+        return true;
+    }
+
+    public function createAuthenticatedToken(UserInterface $user, $providerKey): OAuth2Token
+    {
+        $tokenUser = $user instanceof NullUser ? null : $user;
+
+        $oauth2Token = $this->oauth2TokenFactory->createOAuth2Token($this->psr7Request, $tokenUser, $providerKey);
+
+        if (!$this->isAccessToRouteGranted($oauth2Token)) {
+            throw InsufficientScopesException::create($oauth2Token);
+        }
+
+        $oauth2Token->setAuthenticated(true);
+
+        return $oauth2Token;
+    }
+
+    public function onAuthenticationFailure(Request $request, AuthenticationException $exception): ?Response
+    {
+        $this->psr7Request = null;
+
+        throw $exception;
```

Comment:
Yes, as @froozeify said here https://github.com/trikoder/oauth2-bundle/pull/242#issuecomment-726839920 would be good to implement an extension point to change the response and do what is done here https://github.com/lexik/LexikJWTAuthenticationBundle/blob/2c3448975c5560f4a29ac6f3bfc1fb24673609c0/Security/Guard/JWTTokenAuthenticator.php#L173 (Returning a JsonResponse by default).

The same applies to the `start()` response.

What do you think?

Reply:
I have some more idea, maybe it will be better to wait for this to be merged, and I'll make a merge request after ?

Reply:
Feel free to send the alternative proposal :)

Reply:
There are 3 ways to define a security layer in Symfony: Firewall, Guard Authenticator, and HTTP Authenticator (introduced in SF 5.1).

Introducing extension points using Events it's a good thing, but I think it should be added to every security layer implementation, so I'd prefer to see a different PR with your ideas for the Firewall implementation to see how it works.

I think that this PR should mimic Firewall behavior, so if the Event is introduced before this PR gets merged, I'll add it here.

Reply:
That works for me :)   
I'll try to create the PR this week, nice catch for the Firewall, it should totally have the Events too


@yceruto The other alternative (I will probably add it to a separated PR, didn't check but maybe somebody already asked for it) is to add a setting to configure a Response formatter.  
Currently, when an error occurred the response return is not formatted in any type. Just the message in the response.   
For example for my use case, it will be better that the response is in a JSON format (like lexik/jwt do, so my client can catch the error message).   
I was thinking of a setting where we could define a custom Response format (like `none`, `json` or a user custom response class).

For the moment it's just an idea, but a not a good one, like to know some opinion about it